README.md
## Mental Diffusion

<img style="width:100%;max-width:1280px" src="https://repository-images.githubusercontent.com/646072414/78bd4c5c-feb5-438d-ba36-f6b22f73b7e0">

Stable diffusion headless and web interface<br>
Version 0.1.6 alpha<br>
[Changelog](https://github.com/nimadez/mental-diffusion/blob/main/CHANGELOG)

- [Features](https://github.com/nimadez/mental-diffusion#features)
- [Headless](https://github.com/nimadez/mental-diffusion#headless)
- [Web Interface](https://github.com/nimadez/mental-diffusion#web-interface)
- [Installation](https://github.com/nimadez/mental-diffusion#installation)
- [Models](https://github.com/nimadez/mental-diffusion#models)
- [Known Issues](https://github.com/nimadez/mental-diffusion#known-issues)
- [FAQ](https://github.com/nimadez/mental-diffusion#faq)
- [Credits](https://github.com/nimadez/mental-diffusion#credits)

## Features
- [x] Accelerated Torch 2.0
- [x] Lightweight layer on top of Diffusers
- [x] Fast startup and render
- [x] GPU and CPU *(slower)*
- [x] Easy to use web interface
- [x] Headless console for experts
- [x] Websockets server
- [x] Safetensors-only checkpoints
- [x] JSON configuration file
- [x] Schedulers *(ddpm, ddim, pndm, lms, euler, euler_anc)*
- [x] Text to image
- [x] Image to image
- [x] Image inpainting
- [x] Image outpainting
- [x] Face restoration
- [x] Upscaling 4x
- [x] Batch rendering
- [x] PNG metadata
- [x] Automatic pipeline switch
- [x] File and base64 image inputs *(PNG)*
- [x] Work offline, proxy supported
- [x] No safety checker
- [x] No miners, trackers, and telemetry
- [x] Optimized for affordable hardware
- [x] Optimized for slow internet connections
- [x] Support custom VAE
- [ ] Support LoRA
- [ ] Support ControlNet
- [ ] Support Hypernetwork

## Requirements
- At least 16 GB RAM
- NVIDIA GPU with CUDA compute capability (at least 4 GB memory)

## Headless
<img src="media/console.gif?raw=true">

```
Usage:    headless.py --arg "value"
Example:  headless.py -p "prompt" -out base64
Upscale:  headless.py -sr 0.1 -up true -i "/path-to-image.png" -of true
Metadata: headless.py -meta "/path-to-image.png"
Batch:    headless.py -p "prompt" -st 25 -b 10 -w 1200 -h 400
CKPT:     headless.py -p "prompt" -ck "deliberate_v2"

--help               show this help message and exit
--get        -get    fetch data from server by record id (int)

--batch      -b      number of images to render (def: 1)
--checkpoint -ck     set checkpoint by file name, null = default checkpoint (def: null)
--scheduler  -sc     ddpm, ddim, pndm, lms, euler, euler_anc (def: euler_anc)
--prompt     -p      positive prompt input
--negative   -n      negative prompt input
--width      -w      image width have to be divisible by 8 (def: 512)
--height     -h      image height have to be divisible by 8 (def: 512)
--seed       -s      seed number, 0 to randomize (def: 0)
--steps      -st     steps from 10 to 50, 20-25 is good enough (def: 20)
--guidance   -g      guidance scale, how closely linked to the prompt (def: 7.5)
--strength   -sr     how much respect the final image should pay to the original (def: 0.5)
--image      -i      PNG file path or base64 PNG (def: '')
--mask       -m      PNG file path or base64 PNG (def: '')
--facefix    -ff     true/false, face restoration using gfpgan (def: false)
--upscale    -up     true/false, upscale using real-esrgan 4x (def: false)
--savefile   -sv     true/false, save image to PNG, contain metadata (def: true)
--onefile    -of     true/false, save the final result only (def: false)
--outpath    -o      /path-to-directory (def: ./.temp)
--filename   -f      filename prefix (.png extension is not required)

--metadata   -meta   /path-to-image.png, extract metadata from PNG
--out        -out    stdout 'metadata' or 'base64' (def: metadata)

* When the image is not empty, the pipeline switches to image-to-image
* When the image and mask are not empty, the pipeline switches to inpainting
* Check server.log for previous records
```
<img src="media/badmask.jpg?raw=true"><br>
Incorrect and correct mask image


## Web Interface
> The web interface is a prototype with minimal bugs

<img src="media/webui.jpg?raw=true">

- [x] [Online](https://nimadez.github.io/mental-diffusion/webui) and offline webui
- [x] Drag and drop workflow
- [x] Image comparison A/B<br>
    A - Front canvas *(left)*<br>
    B - Background image *(right)*<br>
- [x] Painting canvas *(brush, line, eraser, mask, color picker)*
- [x] Canvas editor *(flip, hue, saturation, brightness, contrast, sepia, invert)*
- [x] Styles editor *(use predefined keywords, they are included in metadata)*
- [x] Guide the AI using text, brush strokes and color adjustments
- [x] Quick mask painting
- [x] Generates input and mask images for outpainting
- [x] Autosave prompts
- [x] Autosave PNG with metadata
- [x] Metadata pool *(single or multiple PNG import)*
- [x] Bake canvas to image
- [x] Pan and zoom canvas
- [x] Undo/redo for painting tools *(brush, line, eraser, mask)*

> - Your data is safe and can be loaded again as long as "Autosave File" is checked<br>
> - If you want your painting to combine with the image pixels, you need to bake the canvas
> - To create outpainting, set "Outpaint Padding" size, your initial image and mask will be generated for you (set Strength to 1.0)

<img src="media/schedulers.jpg?raw=true"><br>
> Schedulers, using equal steps, steps is not enough for PNDM/LMS

<img src="media/facefix.jpg?raw=true"><br>
> GFPGAN was applied to the LMS rendering above

<img src="media/guide_paint.jpg?raw=true"><br>
> Painting bloods to guide AI with GFPGAN result

<img src="media/inpainting.jpg?raw=true"><br>
> About 10 inpaint renders, top is the original

<img src="media/outpainting.jpg?raw=true"><br>
> Outpaint examples, padding 128 and 256

<img src="media/outpainting_paintball.jpg?raw=true"><br>
> Outpaint padding 128, use inpainting to cleanup errors

#### Mouse controls
| *Key* | *Action* |
| --- | --- |
| Left Button | drag, draw, select |
| Middle Button | Zoom reset |
| Right Button | Pan canvas |
| Wheel | Zoom canvas in/out |

#### Keyboard shortcuts
| *Key* | *Action* |
| --- | --- |
| Space | Toggle metadata pool |
| D | Drag tool |
| B | Brush tool |
| L | Line tool |
| E | Eraser tool |
| M | Mask tool |
| ] | Increase tool size |
| [ | Decrease tool size |
| + | Increase tool opacity |
| - | Decrease tool opacity |
| CTRL + Enter | Render/Generate |
| CTRL + L | Load PNG metadata |
| CTRL + Z | Undo painting |
| CTRL + X | Redo painting |

## Installation
#### [ Automatic Installation ]
```
curl -o md-installer.py https://raw.githubusercontent.com/nimadez/mental-diffusion/main/installer/md-installer.py
```
#### [ Manual Installation ]
- Download [python-3.10.11-embed-amd64.zip](https://www.python.org/ftp/python/3.10.11/python-3.10.11-embed-amd64.zip)
```
curl https://bootstrap.pypa.io/get-pip.py -k --ssl-no-revoke -o get-pip.py
python get-pip.py
python -m pip install torch==2.0.1+cu118 torchvision==0.15.2+cu118 --extra-index-url https://download.pytorch.org/whl/cu118
python -m pip install accelerate==0.20.3
python -m pip install diffusers==0.18.1
python -m pip install transformers==4.30.2
python -m pip install omegaconf==2.3.0
python -m pip install safetensors==0.3.1
python -m pip install realesrgan==0.3.0
python -m pip install gfpgan==1.3.8
python -m pip install websockets==11.0.3

git clone https://github.com/nimadez/mental-diffusion.git
run.bat       -> start server (url: http://localhost:8011)
headless.py   -> use headless if you are familiar with consoles

* edit "config.json" to define model paths
```
#### [ Automatic One-Time Downloads ]
- 200 MB gfpgan weights *(root directory)*
- 1.7 GB openai/clip-vit-large-patch14 *(huggingface cache)*
```
To prevent re-downloading huggingface cache, add HF cache directory to your environment variables
> setx HF_HOME path-to-dir\.cache\huggingface
```

## Models
Checkpoints:<br>
[v1-5-pruned-emaonly.safetensors](https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/main/v1-5-pruned-emaonly.safetensors)<br>
[sd-v1-5-inpainting.ckpt](https://huggingface.co/runwayml/stable-diffusion-inpainting/blob/main/sd-v1-5-inpainting.ckpt)<br>
[Deliberate_v2.safetensors](https://huggingface.co/XpucT/Deliberate/blob/main/Deliberate_v2.safetensors)<br>
[Deliberate-inpainting.safetensors](https://huggingface.co/XpucT/Deliberate/blob/main/Deliberate-inpainting.safetensors)<br>
[Reliberate.safetensors](https://huggingface.co/XpucT/Reliberate/blob/main/Reliberate.safetensors)<br>
[Reliberate-inpainting.safetensors](https://huggingface.co/XpucT/Reliberate/blob/main/Reliberate-inpainting.safetensors)<br>
[dreamlike-diffusion-1.0.safetensors](https://huggingface.co/dreamlike-art/dreamlike-diffusion-1.0/blob/main/dreamlike-diffusion-1.0.safetensors)<br>
[dreamlike-photoreal-2.0.safetensors](https://huggingface.co/dreamlike-art/dreamlike-photoreal-2.0/blob/main/dreamlike-photoreal-2.0.safetensors)<br>
VAE: [vae-ft-mse-840000-ema-pruned.safetensors](https://huggingface.co/stabilityai/sd-vae-ft-mse-original/blob/main/vae-ft-mse-840000-ema-pruned.safetensors) *(optional)*<br>
GFPGAN: [GFPGANv1.4.pth](https://github.com/TencentARC/GFPGAN/releases/download/v1.3.0/GFPGANv1.4.pth) *(required)*<br>
RealESRGAN: [RealESRGAN_x4plus.pth](https://github.com/xinntao/Real-ESRGAN/releases/download/v0.1.0/RealESRGAN_x4plus.pth) *(required)*<br>
- All **.ckpt** checkpoints converted to **.safetensors** *(security)*
- All checkpoints converted to **fp16** *(smaller size, use [prune.py](https://github.com/nimadez/mental-diffusion/tree/main/scripts/prune.py))*
- All inpainting checkpoints must have "inpainting" in their filename
- VAE is optional but recommended for getting optimal results
- Back to the future, SD v1.x only!

<img src="media/backtothefuture.jpg?raw=true">

> - I do not officially support any model
> - Visit [Civitai.com](https://civitai.com/) for more SD 1.5 checkpoints

## Known Issues
```
- Mental Diffusion is offline, if the internet access is interrupted,
if the connection is established, some data will be send and received
when loading the checkpoint. (huggingface tries to compare files)
```

## FAQ
```
How to speed up rendering?
- Do not constantly update the checkpoint, let it be cached and reused
- Open NVIDIA Control Panel, enable "Adaptive" power management mode

Why does it give a connection error when loading the checkpoint?
Use VPN, enable "use_proxy" in config.json, or disable network
connection. (after you have disabled your network connection, you
should not set proxy to 1)

Why not using python venv?
For portable programs with specific python version and many
packages, I prefer to use embeddable python only, because no
other packages are added, it doesnt infuence the performance.

Is SDXL supported?
SDXL requires 12 GB of video memory, it is not currently supported.
SDXL does not appear to be aimed at the open-source community.
```

## History
```
0.1.6 -> back to the roots, major performance gain #1

- Rename project from "undiff" to mental-diffusion
- Mental-diffusion started with "sdkit" and later evolved into diffusers
- Created for my personal use
```

## License
Code released under the [MIT license](https://github.com/nimadez/mental-diffusion/blob/main/LICENSE).

## Credits
- [Hugging Face](https://huggingface.co/)
- [Stability-AI](https://stability.ai/)
- [RunwayML](https://runwayml.com/)
- [PyTorch](https://pytorch.org/)
- [Diffusers](https://github.com/huggingface/diffusers)
- [AUTOMATIC1111](https://github.com/AUTOMATIC1111/stable-diffusion-webui)
- [sdkit](https://github.com/easydiffusion)
- [ComfyUI](https://github.com/comfyanonymous/ComfyUI)
- [GFPGAN](https://github.com/TencentARC/GFPGAN)
- [Real-ESRGAN](https://github.com/xinntao/Real-ESRGAN)
- [prune.py](https://github.com/lopho/stable-diffusion-prune)
- [Civitai](https://civitai.com/)


CHANGELOG
All notable changes to this project will be documented in this file,
UIX improvement and code management is not included in this file.

[0.1.6]
- update diffusers to patch 0.18.2 (important update)
- update python packages
- update installer.py (fix gfpgan and realesrgan errors)
- webui: remove styles in case of duplicates (prevent prompt overflow)

[0.1.5]
- update diffusers to patch 0.18.1
- CPU is now supported and working
- core revision, faster startup time and webui access
- major performance gain
- no more preloads, everything is built as you progress
- cfg renamed to guidance (guidance scale)
- config.json updated
- add proxy support for restricted countries

[0.1.4]
- update diffusers to 0.18.0
- new styles
- revert mask painting, mask is a tool now
- mask image supports drag and drop
- upscaled image save to file only (not returned to webui)
- code cleanup

[0.1.3]
- massive webui update, improve the code and uix
- webui is now available online and offline
- fix md unable to load new checkpoint
- md scan directory for safetensors, config.json updated
- improve readability of metadata pool tags
- add render pipeline to PNG metadata
- add batch rendering to both webui and headless
- rewrite and improve image drag and drop

[0.1.2]
- add brush softness
- add keyboard shortcuts
- canvas pan and zoom are now global
- inpaint rework, mask painting is a lot easier, just mask and render
- webui pipeline revert to manual (in case you want to keep the mask but do img2img)
- a lot of webui bug fixes and minor improvements, shaping up the base
- add option to save the one last file only
- overall workflow improvements

[0.1.1]
- debugging and optimization
- add mouse wheel zoom
- add server max_size limit, fix upscale-base64 error
- limit width and height to divisible by 8
- add custom default checkpoint to config.json
- installer preserve user config.json (create backup before update)
- automatic canvas resize on image drop and load
- config.json checkpoints are fully optional now
- new outpainting, enjoy!
- new smooth brushes
- new undo/redo for painting only

[0.1.0]
- update diffusers to 0.17.1
- fix inpainting checkpoint loader
- add metadata pool
- add server connection status
- add line drawing tool
- add canvas editor
- debugging web interface

[0.0.9]
- electron removed
- you can directly use http://localhost:8011 to load the webui
- server updated to websockets (remember to force-check python packages)
- add support for more schedulers
- update and fixed python requirements
- fix the diffusers StableDiffusionPipeline.from_ckpt() not loading checkpoints when offline
https://github.com/huggingface/diffusers/issues/3729

[0.0.8]
- update to diffusers 0.17.0 pip release
- fix resizable interface
- add python installer script
- finish manual installation readme
- add metadata extraction to headless
- sort and arrange headless args
- add multiple options to config.json
- remove base64 data from server.log
- fix seed rng to count more possible values
- respect existing filenames
- reduce server memory overhead
- free up 2gb space by eliminating safety_checker (HF cache)
- speed up the core and startup time


src/md.py
#
# mental-diffusion's core
#
import logging
log = logging.getLogger("mental-diffusion")

import os
import sys
import json
from gc import collect

from numpy import array as np_array
from numpy import uint8 as np_uint8
from PIL.PngImagePlugin import PngInfo
from PIL import Image

from torch import (
    cuda,
    float16,
    Generator
)

from diffusers.models.attention_processor import AttnProcessor2_0
from diffusers import (
    StableDiffusionPipeline,
    StableDiffusionImg2ImgPipeline,
    StableDiffusionInpaintPipeline,
    DDPMScheduler,
    DDIMScheduler,
    PNDMScheduler,
    LMSDiscreteScheduler,
    EulerDiscreteScheduler,
    EulerAncestralDiscreteScheduler
)

from gfpgan import GFPGANer
from realesrgan import RealESRGANer
from basicsr.archs.rrdbnet_arch import RRDBNet

import utils


class Context():
    def __init__(self):
        self.version = "0.1.6"
        
        self.use_CPU = 0
        self.use_VAE = 0

        self.device = "cpu"
        self.dtype = None

        self.configs = []
        self.paths = {}
        self.checkpoints = []

        self.checkpoint_name = ""
        self.scheduler_name = ""

        self.pipe = None
        self.pipe_name = ""

        self.gfpgan = None
        self.realesrgan = None

ctx = Context()


def preload(config):
    clear_cache()

    ctx.use_CPU = config["use_CPU"]
    ctx.use_VAE = config["use_VAE"]
    ctx.checkpoint_name = config["checkpoint"]
    ctx.scheduler_name = "euler_anc" # ddpm, ddim, pndm, lms, euler, euler_anc

    ctx.configs = [
        "configs/v1-inference.yaml",
        "configs/v1-inpainting-inference.yaml",
        "configs/x4-upscaling.yaml"
    ]

    ctx.paths = {
        "checkpoints_root": config["checkpoints_root"],
        "vae": config["vae"],
        "gfpgan": config["gfpgan"],
        "realesrgan": config["realesrgan"]
    }

    if not os.path.exists(config["checkpoints_root"]):
        print("No checkpoint directory found, unable to start the mental-diffusion.\nHave you set up the checkpoints_root in config.json file?")
        sys.exit(0)

    for ckpt in os.listdir(config["checkpoints_root"]):
        if ckpt.endswith(".safetensors"):
            ctx.checkpoints.append(os.path.splitext( os.path.basename(ckpt) )[0])

    if len(ctx.checkpoints) == 0:
        print("No checkpoints found, unable to start the mental-diffusion.\nHave you set up the checkpoints path in config.json file?")
        sys.exit(0)

    if not os.path.exists(ctx.paths["checkpoints_root"] + ctx.checkpoint_name + ".safetensors"):
        print("Default checkpoint does not exist, unable to start the mental-diffusion.\nHave you set up the default checkpoint in config.json file?")
        sys.exit(0)

    if ctx.use_VAE == 1 and not os.path.exists(ctx.paths["vae"]):
        print("Custom VAE is enabled, but no VAE found, unable to start the mental-diffusion.\nHave you set up the VAE in config.json file?")
        sys.exit(0)
        
    if not os.path.exists(ctx.paths["gfpgan"]):
        print("No GFPGAN found, unable to start the mental-diffusion.\nHave you set up the GFPGAN in config.json file?")
        sys.exit(0)

    if not os.path.exists(ctx.paths["realesrgan"]):
        print("No Real-ESRGAN found, unable to start the mental-diffusion.\nHave you set up the Real-ESRGAN in config.json file?")
        sys.exit(0)

    device_setup()


def init():
    log.info("initialized.")


def device_setup():
    ctx.dtype = float16
    ctx.device = "cuda" if cuda.is_available() else "cpu"
    if ctx.use_CPU == 1 or ctx.device == "cpu":
        ctx.device = "cpu"
        ctx.dtype = None


def load_checkpoint(name):
    if name == "null": # to use the current checkpoint (for headless)
        name = ctx.checkpoint_name

    if ctx.pipe and ctx.checkpoint_name == name:
        return # update checkpoint or use the current

    log.info(f"loading checkpoint [{ name }] ...")

    if not "inpainting" in name.lower():
        cfg = ctx.configs[0] # v1-inference
    else:
        cfg = ctx.configs[1] # v1-inpainting-inference

    logging.getLogger("diffusers").setLevel(logging.ERROR) # hide safety message
    sys.stdout = open(os.devnull, 'w') # remove "global_step key not found" warning
    ctx.checkpoint_name = name
    ctx.pipe = utils.pipe_from_ckpt(
        StableDiffusionPipeline,
        ctx.paths["checkpoints_root"] + name + ".safetensors",
        original_config_file = cfg,
        local_files_only = True,
        use_safetensors = True,
        torch_dtype = ctx.dtype,
        prediction_type = "epsilon",    # SD 1.x
        image_size = 512,               # SD 1.x
        force_download = False,
        resume_download = True,
        load_safety_checker = False)
    ctx.pipe.requires_safety_checker = False
    ctx.pipe.safety_checker = None
    ctx.pipe.to(ctx.device, ctx.dtype)
    sys.stdout = sys.__stdout__

    load_vae(cfg)
    pipe_optimizer()
    log.info("checkpoint ready.")


def load_vae(config):
    if ctx.use_VAE == 0: return

    ctx.pipe.vae = utils.load_vae_weights(ctx.paths["vae"], ctx.device, 512, config)
    if ctx.pipe.vae is not None:
        ctx.pipe.vae.to(ctx.device, ctx.dtype)
        log.info("vae loaded")
    else:
        log.error("unable to load vae")


def pipe_optimizer():
    ctx.pipe.unet.set_attn_processor(AttnProcessor2_0())
    ctx.pipe.vae.decoder.mid_block.attentions[0]._use_2_0_attn = True
    if ctx.device != "cpu":
        ctx.pipe.enable_model_cpu_offload()     # memory optimization
        ctx.pipe.enable_attention_slicing(1)    # low vram usage (auto|max|8)
        ctx.pipe.enable_vae_slicing()           # sliced VAE decode for larger batches
        ctx.pipe.enable_vae_tiling()            # tiled VAE decode/encode for large images


def set_scheduler(name):
    match name:
        case "ddpm":
            ctx.pipe.scheduler = DDPMScheduler.from_config(ctx.pipe.scheduler.config)
        case "ddim":
            ctx.pipe.scheduler = DDIMScheduler.from_config(ctx.pipe.scheduler.config)
            ctx.pipe.scheduler.register_to_config(clip_sample=False)
        case "pndm":
            ctx.pipe.scheduler = PNDMScheduler.from_config(ctx.pipe.scheduler.config)
            ctx.pipe.scheduler.register_to_config(skip_prk_steps=True)
        case "lms":
            ctx.pipe.scheduler = LMSDiscreteScheduler.from_config(ctx.pipe.scheduler.config)
        case "euler":
            ctx.pipe.scheduler = EulerDiscreteScheduler.from_config(ctx.pipe.scheduler.config)
        case "euler_anc":
            ctx.pipe.scheduler = EulerAncestralDiscreteScheduler.from_config(ctx.pipe.scheduler.config)
    ctx.scheduler_name = name
    log.info(f"scheduler [{name}]")


def inference_stablediffusion(
    prompt, negative,
    width, height,
    seed, steps, guidance, strength,
    initimage, maskimage):

    generator = Generator(ctx.device).manual_seed(seed)
    log.info(f"seed [{ seed }]")

    img = None

    if initimage == "":
        ctx.pipe_name = "txt2img"
        log.info("pipeline [txt2img] ...")
        img = ctx.pipe(
            prompt = prompt,
            width = width,
            height = height,
            num_inference_steps = steps,
            guidance_scale = guidance,
            negative_prompt = negative,
            num_images_per_prompt = 1,
            eta = 0.0, # only applies to DDIMScheduler
            generator = generator,
            latents = None,
            prompt_embeds = None,
            negative_prompt_embeds = None,
            output_type = "pil", # "str"
            return_dict = True,
            callback = None,
            callback_steps = 1,
            cross_attention_kwargs = None,
            guidance_rescale = 0.7)[0][0] #def: 0.7
    elif initimage != "" and maskimage == "":
        ctx.pipe_name = "img2img"
        log.info("pipeline [img2img] ...")
        img = StableDiffusionImg2ImgPipeline(
            vae = ctx.pipe.vae,
            text_encoder = ctx.pipe.text_encoder,
            tokenizer = ctx.pipe.tokenizer,
            unet = ctx.pipe.unet,
            scheduler = ctx.pipe.scheduler,
            safety_checker = None,
            feature_extractor = None,
            requires_safety_checker = False,
            )(
                prompt = prompt,
                image = initimage,
                strength = strength,
                num_inference_steps = steps,
                guidance_scale = guidance,
                negative_prompt = negative,
                num_images_per_prompt = 1,
                eta = 0.0,
                generator = generator,
                prompt_embeds = None,
                negative_prompt_embeds = None,
                output_type = "pil",
                return_dict = True,
                callback = None,
                callback_steps = 1,
                cross_attention_kwargs = None)[0][0]
    elif initimage != "" and maskimage != "":
        ctx.pipe_name = "inpaint"
        log.info("pipeline [inpaint] ...")
        img = StableDiffusionInpaintPipeline(
            vae = ctx.pipe.vae,
            text_encoder = ctx.pipe.text_encoder,
            tokenizer = ctx.pipe.tokenizer,
            unet = ctx.pipe.unet,
            scheduler = ctx.pipe.scheduler,
            safety_checker = None,
            feature_extractor = None,
            requires_safety_checker = False,
            )(
                prompt = prompt,
                image = initimage,
                mask_image = maskimage,
                width = width,
                height = height,
                strength = strength,
                num_inference_steps = steps,
                guidance_scale = guidance,
                negative_prompt = negative,
                num_images_per_prompt = 1,
                eta = 0.0,
                generator = generator,
                latents = None,
                prompt_embeds = None,
                negative_prompt_embeds = None,
                output_type = "pil",
                return_dict = True,
                callback = None,
                callback_steps = 1,
                cross_attention_kwargs = None)[0][0]

    del initimage
    del maskimage
    del generator
    log.info("image generated.")
    return img


def inference_gfpgan(image, width=512, height=512, upscale=2):
    image = image.resize((width, height))
    image = image.convert("RGB")
    image = np_array(image, dtype=np_uint8)[..., ::-1]

    if ctx.gfpgan == None:
        ctx.gfpgan = GFPGANer(
            model_path = ctx.paths["gfpgan"],
            upscale = 1, # def: 2
            arch = "clean",
            channel_multiplier = 2,
            bg_upsampler = None)
        
    _, _, img = ctx.gfpgan.enhance(
            image,
            has_aligned = False,
            only_center_face = False,
            paste_back = True,
            weight = 0.5)

    img = img[:, :, ::-1]
    img = Image.fromarray(img)
    del image
    return img


def inference_realesrgan(image, width=512, height=512, scale=4):
    image = image.resize((width, height))
    image = image.convert("RGB")
    image = np_array(image, dtype=np_uint8)[..., ::-1]

    if ctx.realesrgan == None:
        ctx.realesrgan = RealESRGANer(
            scale = 4, # def netscale
            model_path = ctx.paths["realesrgan"],
            dni_weight = None,
            model = RRDBNet(num_in_ch=3, num_out_ch=3, num_feat=64, num_block=23, num_grow_ch=32, scale=4), # (x4 RRDBNet model)
            tile = 256, # ~512
            tile_pad = 10,
            pre_pad = 0,
            half = "fp16",
            gpu_id = 0)

    img, _ = ctx.realesrgan.enhance(image, outscale=scale)

    img = img[:, :, ::-1]
    img = Image.fromarray(img)
    del image
    return img


def prepare_input_image(img, width, height, label):
    if img.startswith('data:image/png'):
        img = utils.base64Decode(img)
        log.info(f"{ label } image [base64]")
    elif img.endswith('.png') or img.endswith('.PNG'): # filepath
        log.info(f"{ label } image [{ img }]")
        img = Image.open(img)
    else:
        log.error(f"bad { label } image [{ img[:50] }]")
        return None
    img = img.resize((width, height))
    return img.convert("RGB")


def create_image(
    checkpoint, scheduler,
    prompt, negative,
    width, height,
    seed, steps, guidance, strength,
    initimage, maskimage,
    facefix, upscale,
    savefile, onefile, outpath, filename):

    load_checkpoint(checkpoint)
    set_scheduler(scheduler)

    if initimage:
        initimage = prepare_input_image(initimage, int(width), int(height), "init")

    if maskimage:
        maskimage = prepare_input_image(maskimage, int(width), int(height), "mask")
            
    img = inference_stablediffusion(
        prompt, negative,
        int(width), int(height),
        int(seed), int(steps), float(guidance), float(strength),
        initimage, maskimage)

    del initimage
    del maskimage

    metadict = {
        "version": f"MD { ctx.version }",
        "checkpoint": ctx.checkpoint_name,
        "scheduler": ctx.scheduler_name,
        "pipeline": ctx.pipe_name,
        "prompt": prompt,
        "negative": negative,
        "width": width,
        "height": height,
        "seed": seed,
        "steps": steps,
        "guidance": guidance,
        "strength": strength,
        "image": "null",
        "mask": "null",
        "facefix": facefix,
        "upscale": upscale,
        "savefile": savefile,
        "onefile": onefile,
        "outpath": outpath,
        "filename": filename
    }
    metadata = PngInfo()
    metadata.add_text("MD", json.dumps(metadict))

    count = 1
    savepath = outpath + "/" + filename
    while os.path.exists(savepath + ".png"): # respect existing filenames
        savepath = outpath + "/" + filename + "_" + str(count)
        count += 1

    if savefile and not onefile:
        img.save(f"{ savepath }.png", pnginfo=metadata)
        log.info(f"image saved to { savepath }.png")

    if facefix:
        img = inference_gfpgan(img, width=width, height=height, upscale=1)
        log.info("filter [gfpgan] applied.")
        if savefile and not onefile:
            img.save(f"{ savepath }_ff.png", pnginfo=metadata)
            log.info(f"image saved to { savepath }_ff.png")

    out_base64 = utils.base64Encode(img, "png")

    if upscale:
        img = inference_realesrgan(img, width=width, height=height, scale=4)
        log.info("filter [realesrgan 4x] applied.")
        if savefile and not onefile:
            img.save(f"{ savepath }_4x.png", pnginfo=metadata)
            log.info(f"image saved to { savepath }_4x.png")

    if savefile and onefile:
        img.save(f"{ savepath }.png", pnginfo=metadata)
        log.info(f"image saved to { savepath }.png")

    del img
    del metadata
    clear_cache()
    log.info("done")
    return [ metadict, out_base64 ]


def clear_cache():
    collect()
    if cuda.is_available():
        cuda.empty_cache()
        cuda.ipc_collect()

def clear_pipe():
    if ctx.pipe:
        del ctx.pipe
    del ctx.gfpgan
    del ctx.realesrgan
    clear_cache()
